apps/hash-graph/lib/graph/src/shared/identifier.rs

```diff
@@ -176,7 +176,46 @@ impl IntoIterator for EntityIdWithIntervalSet {
 
     fn into_iter(self) -> Self::IntoIter {
         self.inner.into_iter().flat_map(|(entity_id, intervals)| {
+            // This merges overlapping intervals
+            //  Examples   |       1       |       2       |       3
+            //  ===========|===============|===============|===============
+            //  Interval A | [--]          | (--]          | [--]
+            //  Interval B |      [-]      |      [-)      |      [-)
+            //  Interval C |  [--]         |  [--]         |  [---]
+            //  Interval D |           [-] |           [-] |        [----]
+            //  -----------|---------------|---------------|---------------
+            //  Union      | [------]  [-] | (------)  [-] | [-----------]
+
+            // TODO: This is a very primitive implementation. Instead of iterating over all elements
```

Comment:
I tested the generator-approach and it's also fairly straight forward, but I didn't use it here because the generator API is pretty unstable. It's also possible to implement a custom iterator, but I considered that not worth the effort. We can revisit this if it turns out, that the intermediate vector is a problem. (Keeping in mind, that this requires _at least one_ allocation per entity id used as right endpoint)